client/tinychain/ml/linalg.py

```diff
@@ -273,3 +277,251 @@ def svd(matrix: Tensor) -> Tuple:
+            cxt.g_final = After(While(cond, step, Map(j=l)), cxt.g_init)
+
+            cxt.rescale = After(cxt.s_final, cxt.row_i.write(cxt.row_i * scale))
+            return After(cxt.rescale, cxt.g_final)
+
+        cxt.update_cols = update_cols
+        cxt.update_rows = update_rows
+
+        cxt.update_e = e[i].write(scale * g)
+        cxt.l = UInt(After(cxt.update_e, i + 1))
+        cxt.scale_init = Float(If(i < txn.m, dot(U[i:, i], U[i:, i]), scale))
+        cxt.g_init = Float(If(
+            Bool(i < txn.m).logical_and(cxt.scale_init.abs() > EPS),
+            cxt.update_cols(i=i, l=cxt.l, scale=cxt.scale_init),
+            0.0))
+
+        cxt.scale_new = Float(If(
+            Bool(i < txn.m).logical_and(i != txn.n - 1),
+            dot(U[i, cxt.l:], U[i, cxt.l:]),
+            scale))
+
+        cxt.scale_final = Float(After(W[i].write(cxt.scale_init * cxt.g_init), cxt.scale_new))
+
+        g = If(
+            Bool(i < txn.m).logical_and(i != txn.n - 1).logical_and(cxt.scale_init > EPS),
+            cxt.update_rows(l=cxt.l, scale=cxt.scale_final),
+            0.0)
+
+        return Map(i=i + 1, scale=cxt.scale_final, g=g, U=U, W=W, e=e)
+
+    @closure(txn.n)
+    @post_op
+    def cond(i: UInt) -> Bool:
+        return i < txn.n
+
+    return Map(While(cond, step, Map(i=0, scale=0.0, g=0.0, U=U, W=W, e=e)))
+
+
+# TODO: make it work like numpy version
+@post_op
+def rht(txn, U: Tensor, V: Tensor, e: Tensor) -> Map:
+    """Accumulation of right hand transformations (rht)"""
+
+    txn.shape = U.shape
+    txn.m, txn.n = [UInt(dim) for dim in txn.shape.unpack(2)]
+    
+    @closure(U, V, e, txn.m, txn.n)
+    @post_op
+    def step(cxt, i: UInt, l: UInt, g: F32) -> Map:
+        
+        @closure(U, V, e, txn.m, txn.n)
+        @post_op
+        def update_cols_wo_last(cxt, i: UInt, l: UInt, g: F32) -> F32:
+
+            @closure(U, V, e, txn.m, txn.n)
+            @post_op
+            def update_cols(cxt, i: UInt, l: UInt, g: F32) -> F32:
+                
+                @closure(U, V, e, txn.m, txn.n)
+                @post_op
+                def step(cxt, j: UInt, l: UInt, i: UInt) -> Map:
+                    cxt.col_j = V[l:, j]
+                    cxt.update_col = cxt.col_j.write(cxt.col_j + V[l:, i] * dot(U[i, l:], cxt.col_j))
+                    return Map(After(
+                        cxt.update_col,
+                        Map(j=j + 1, n=txn.n, l=l, i=i)
+                    ))
+
+                @closure(txn.n)
+                @post_op
+                def cond(j: UInt) -> Bool:
+                    return j < txn.n
+
+                loop = After(V[l:,i].write((U[i,l:] / U[i,l]) / g), While(cond, step, Map(j=l, n=txn.n, l=l, i=i)))
+
+                return loop
+
+            cxt.update_cols = update_cols
+            cxt.res = cxt.update_cols(n=txn.n, l=l, i=i)
+            return If((g != 0.0), cxt.res)
+
+        cxt.update_g = Float(e[i].copy())
```

Comment:
`e[i]` is a `Number` so it doesn't have a copy method (that's why you're getting "invalid value for tensor: 0"--`copy` calls `Tensor.copy_from` and it complains that it can't copy from `0`)